File: src/opentrials/adapters/osp/inspect_model.py
"""Read-only OSP adapter for generic PKML model discovery.

Deliberately separate from ``OspSimulationEngine`` and
``observed_data.py``: this never executes a simulation and never
interprets what it finds as a verified OpenTrials capability -- it
reports exactly what ``ospsuite`` itself can discover about a model file
(molecules, administration event containers and their parameter paths,
candidate output paths, a mutable-parameter count, a population-support
heuristic), and nothing more. Turning a discovery into a registered
``ModelCapabilityProfile`` is a separate, explicitly researcher-reviewed
step (see ``sdk.model_onboarding``).
"""

from __future__ import annotations

import json
import os
import subprocess
import tempfile
from collections.abc import Mapping
from pathlib import Path
from typing import Any

from opentrials.adapters.osp.engine import DEFAULT_DOTNET_ROOT, DEFAULT_FRAMEWORK_RSCRIPT
from opentrials.core.serialization import SchemaDocument, document

MODEL_INSPECTION_WORKER_REQUEST_SCHEMA = "opentrials.osp.model-inspection-worker-request"
MODEL_INSPECTION_WORKER_RESPONSE_SCHEMA = "opentrials.osp.model-inspection-worker-response"
MODEL_INSPECTION_WORKER_SCHEMA_VERSION = "1.0.0"


class OspModelInspectionWorkerError(RuntimeError):
    """Raised when the model-inspection PKML worker cannot produce a valid result."""


def inspect_model_pkml(
    pkml_path: Path,
    *,
    r_libs_user: str | None = None,
    rscript_path: Path = DEFAULT_FRAMEWORK_RSCRIPT,
    worker_path: Path | None = None,
    dotnet_root: str = DEFAULT_DOTNET_ROOT,
    timeout_seconds: float = 120.0,
) -> Mapping[str, Any]:
    """Ask OSP to discover a PKML file's own structure and return its raw report.

    Returns the worker's own JSON payload unmodified -- this function
    performs no interpretation or classification beyond what the worker
    itself already reports; see ``sdk.model_onboarding`` for the
    researcher-facing scaffold that turns this into something reviewable.
    """
    worker = worker_path or Path(__file__).with_name("inspect_model.R")
    if not rscript_path.is_file():
        raise OspModelInspectionWorkerError(f"Rscript executable does not exist: {rscript_path}")
    if not worker.is_file():
        raise OspModelInspectionWorkerError(f"OSP worker script does not exist: {worker}")

    request = document(
        MODEL_INSPECTION_WORKER_REQUEST_SCHEMA,
        {"pkml_path": str(pkml_path)},
        MODEL_INSPECTION_WORKER_SCHEMA_VERSION,
    )

    environment = os.environ.copy()
    environment["DOTNET_ROOT"] = dotnet_root
    if r_libs_user is not None:
        environment["R_LIBS_USER"] = r_libs_user

    with tempfile.TemporaryDirectory(prefix="opentrials-osp-inspect-") as temporary_directory:
        temporary_path = Path(temporary_directory)
        request_path = temporary_path / "request.json"
        response_path = temporary_path / "response.json"
        request_path.write_text(request.canonical_json(), encoding="utf-8")
        completed = subprocess.run(
            [
                str(rscript_path),
                str(worker),
                "--input",
                str(request_path),
                "--output",
                str(response_path),
            ],
            capture_output=True,
            check=False,
            env=environment,
            text=True,
            timeout=timeout_seconds,
        )
        if not response_path.is_file():
            raise OspModelInspectionWorkerError(
                "OSP model-inspection worker completed without writing a response artifact: "
                f"{completed.stderr.strip() or completed.stdout.strip()}"
            )
        try:
            decoded: Mapping[str, Any] = json.loads(response_path.read_text(encoding="utf-8"))
            response = SchemaDocument.model_validate(decoded)
        except (OSError, json.JSONDecodeError, ValueError) as error:
            raise OspModelInspectionWorkerError(
                f"OSP model-inspection worker emitted an invalid JSON response: {error}"
            ) from error
        if response.schema_id != MODEL_INSPECTION_WORKER_RESPONSE_SCHEMA:
            raise OspModelInspectionWorkerError(
                f"Unexpected model-inspection worker response schema: {response.schema_id!r}"
            )
        if completed.returncode != 0 or response.payload.get("status") != "SUCCEEDED":
            raise OspModelInspectionWorkerError(
                "OSP model-inspection worker did not succeed: "
                f"{response.payload.get('error', completed.stderr.strip())}"
            )
        return response.payload

File: src/opentrials/adapters/osp/inspect_model.R
#!/usr/bin/env Rscript

# Read-only, generic PKML inspection worker for `opentrials model inspect`.
# Reports what ospsuite itself can discover about a simulation file --
# molecule names, administration event containers and their mutable
# parameter paths, candidate observable output paths, a total mutable
# parameter count, and a population-compatibility heuristic. Discovery
# only: this worker makes no claim about which of these facts constitute
# a verified, usable OpenTrials capability -- that judgment belongs to a
# researcher reviewing `opentrials model init`'s generated scaffold, never
# to this script.

worker_request_schema <- "opentrials.osp.model-inspection-worker-request"
worker_response_schema <- "opentrials.osp.model-inspection-worker-response"
worker_schema_version <- "1.0.0"

arguments <- commandArgs(trailingOnly = TRUE)
argument_value <- function(flag) {
  index <- match(flag, arguments)
  if (is.na(index) || index == length(arguments)) {
    stop(sprintf("Missing required argument %s", flag), call. = FALSE)
  }
  arguments[[index + 1]]
}

input_path <- argument_value("--input")
output_path <- argument_value("--output")

write_response <- function(payload) {
  response <- list(
    schema = worker_response_schema,
    schema_version = worker_schema_version,
    payload = payload
  )
  jsonlite::write_json(
    response, output_path, auto_unbox = TRUE, pretty = FALSE, na = "null", digits = NA
  )
}

# Candidate administration-parameter roles this project has previously
# verified matter (see models/profiles/aciclovir_iv.py's own dose/
# administration-time/infusion-duration paths, discovered this same way).
# Any other leaf name is still reported, just not classified into a role.
ROLE_SUFFIXES <- list(
  dose = "Dose",
  start_time = "Start time",
  infusion_duration = "Infusion time"
)

run_worker <- function() {
  request <- jsonlite::fromJSON(input_path, simplifyVector = FALSE)
  if (!identical(request$schema, worker_request_schema)) {
    stop("Unsupported model inspection worker request schema.", call. = FALSE)
  }
  if (!identical(request$schema_version, worker_schema_version)) {
    stop("Unsupported model inspection worker request schema version.", call. = FALSE)
  }
  payload <- request$payload
  if (!is.character(payload$pkml_path) || !nzchar(payload$pkml_path)) {
    stop("Worker request requires a non-empty pkml_path.", call. = FALSE)
  }
  if (!file.exists(payload$pkml_path)) {
    stop(sprintf("PKML file does not exist: %s", payload$pkml_path), call. = FALSE)
  }

  suppressPackageStartupMessages(library(ospsuite))
  simulation <- loadSimulation(payload$pkml_path)

  all_paths <- getAllParameterPathsIn(simulation)
  molecule_paths <- tryCatch(getAllMoleculePathsIn(simulation), error = function(e) character(0))
  molecule_names <- unique(vapply(
    strsplit(molecule_paths, "|", fixed = TRUE),
    function(parts) parts[[length(parts)]],
    character(1)
  ))

  event_paths <- Filter(function(p) startsWith(p, "Events|"), all_paths)
  container_names <- unique(sub("^(Events\\|[^|]+\\|).*$", "\\1", event_paths))
  administrations <- lapply(container_names, function(container) {
    own_paths <- Filter(function(p) startsWith(p, container), event_paths)
    leaf_names <- vapply(
      strsplit(own_paths, "|", fixed = TRUE),
      function(parts) parts[[length(parts)]],
      character(1)
    )
    roles <- list()
    for (role in names(ROLE_SUFFIXES)) {
      matches <- own_paths[leaf_names == ROLE_SUFFIXES[[role]]]
      if (length(matches) > 0) {
        roles[[role]] <- matches[[1]]
      }
    }
    list(container = container, parameter_paths = own_paths, roles = roles)
  })

  output_paths <- tryCatch(getAllObserverPathsIn(simulation), error = function(e) character(0))
  # Administration-internal bookkeeping observers (dose-container running
  # totals) are not candidate PK outputs -- excluded so this list stays
  # focused on organism-level observable quantities.
  output_paths <- Filter(function(p) !startsWith(p, "Events|"), output_paths)

  population_support <- any(grepl("^Organism\\|Age$", all_paths))

  write_response(list(
    status = "SUCCEEDED",
    generated_at = format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ", tz = "UTC"),
    r_version = R.version.string,
    ospsuite_version = as.character(utils::packageVersion("ospsuite")),
    name = simulation$name,
    molecule_names = as.list(molecule_names),
    administrations = administrations,
    output_paths = as.list(output_paths),
    mutable_parameter_count = length(all_paths),
    population_support_detected = population_support
  ))
}

tryCatch(
  run_worker(),
  error = function(error) {
    write_response(list(
      status = "FAILED",
      error = conditionMessage(error),
      generated_at = format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ", tz = "UTC")
    ))
    quit(status = 1)
  }
)
